### ISSUE TITLE
Rare Exception - java.lang.InternalError: Thread starting during runtime shutdown

### ISSUE THREAD

**Issue by jkasten2** [OneSignal/OneSignal-Android-SDK#287]
Rare crash on Android 5 to 7 devices on SDK version 3.5.8. Was reported to OneSignal support, haven't been able to reproduce yet.

```Java
java.lang.InternalError: Thread starting during runtime shutdown
at java.lang.Thread.nativeCreate(Native Method)
at java.lang.Thread.start(Thread.java:1063)
at com.onesignal.OneSignalRestClient.makeRequest(OneSignalRestClient.java:98)
at com.onesignal.OneSignalRestClient.postSync(OneSignalRestClient.java:86)
at com.onesignal.OneSignalStateSynchronizer.syncUserState(OneSignalStateSynchronizer.java:616)
at com.onesignal.OneSignalStateSynchronizer$NetworkHandlerThread$1.run(OneSignalStateSynchronizer.java:532)
at android.os.Handler.handleCallback(Handler.java:739)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:148)
at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by jkasten2** [comment]
Another stack from issue #329 

```Java
Fatal Exception: java.lang.InternalError: Thread starting during runtime shutdown
       at java.lang.Thread.nativeCreate(Thread.java)
       at java.lang.Thread.start(Thread.java:1063)
       at com.onesignal.OneSignalRestClient.putSync(Unknown Source)
       at com.onesignal.OneSignalRestClient.postSync(Unknown Source)
       at com.onesignal.OneSignalStateSynchronizer.handleJsonArray(Unknown Source)
       at com.onesignal.OneSignalStateSynchronizer$NetworkHandlerThread$1.run(Unknown Source)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:148)
       at android.os.HandlerThread.run(HandlerThread.java:61)

#0. Crashed: OSH_NetworkHandlerThread: 0 0 0x0000000000000000
       at java.lang.Thread.nativeCreate(Thread.java)
       at java.lang.Thread.start(Thread.java:1063)
       at com.onesignal.OneSignalRestClient.putSync(Unknown Source)
       at com.onesignal.OneSignalRestClient.postSync(Unknown Source)
       at com.onesignal.OneSignalStateSynchronizer.handleJsonArray(Unknown Source)
       at com.onesignal.OneSignalStateSynchronizer$NetworkHandlerThread$1.run(Unknown Source)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:148)
       at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by ishratkhan** [comment]
Same problem at

https://github.com/OneSignal/OneSignal-Android-SDK/blob/274f98b0a0a37d79422f7eacdc102358bbbd57e1/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalPrefs.java#L105

**Comment by jkasten2** [comment]
@ishratkhan The specific instance you linked is fixed in PR #479 and will go out in the 3.8.3 release

However I am still unsure how your app or others are getting into a Java runtime shutting down state to reproduce these issues. We plan to move more Java Thread creations into `HandlerThreads` as an optimization. This will fix the `Thread starting during runtime shutdown` errors as we do so but I don't believe we will be able to eliminate all `new Thread()` calls durning the usage of the app.

If you have any more details on how your app would be getting it this shutdown state it would be very helpful. Thanks!

**Comment by rgomezp** [comment]
Closing issue. Note: request still under consideration
